File: neetcode150/medium/36-Valid-Sudoku.py
#!/usr/bin/python3
from collections import defaultdict

class Solution:
  # Optimal solution O(n^2)
  def isValidSudoku(self, board) -> bool:
    row = defaultdict(set)
    col = defaultdict(set)
    grid = defaultdict(set)
    for r in range(9):
      for c in range(9):
        if board[r][c] == ".":
          continue
        if (board[r][c] in row[r] or 
          board[r][c] in col[c] or
          board[r][c] in grid[(r // 3, c // 3)]):
          return False
        row[r].add(board[r][c])
        col[c].add(board[r][c])
        grid[(r // 3, c // 3)].add(board[r][c])
    return True

File: neetcode150/medium/36-Valid-Sudoku.cpp
#include <vector>
#include <map>
#include <set>

using namespace std;

class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        map<pair<int, int>, set<char>> squares({});
        map<int, set<char>> col({});
        map<int, set<char>> row({});
        for (int i = 0; i < 9; i++){
            for (int j = 0; j < 9; j++){
                char v = board[i][j];
                if (v == '.') {
                    continue;
                }
                pair<int, int> sqr_key = {i/3, j/3};
                set<char>& sqr_set = squares[sqr_key];
                auto a = sqr_set.insert(v);
                auto b = row[i].insert(v);
                auto c = col[j].insert(v);
                if (!a.second || !b.second || !c.second) {
                    return false;
                }
            }
        }    
        return true;
    }
};